hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/BatchStage.java

```diff
@@ -74,6 +80,45 @@
             @Nonnull DistributedBiFunction<? super C, ? super T, ? extends Traverser<? extends R>> flatMapFn
     );
 
+    /**
+     * Attaches to this stage a stage that extracts a key from each item
+     * (using the supplied {@code keyFn}) and emits just the items that
+     * are distinct according to that key (no two emitted items map to the
+     * same key). There is no guarantee among the items with the same key
+     * which one it will emit. Returns the newly attached stage.
+     *
+     * @param keyFn function that extracts the key from the item
+     * @param <K> type of the key
+     * @return the newly attached stage
+     */
+    @Nonnull
+    <K> BatchStage<T> distinctBy(DistributedFunction<? super T, ? extends K> keyFn);
+
+    /**
+     * Attaches to this stage a stage that emits just the items that are
+     * distinct according to their definition of equality ({@code equals}
+     * and {@code hashCode}). There is no guarantee among the items with the
+     * same key which one it will emit. Returns the newly attached stage.
+     *
+     * @return the newly attached stage
+     */
+    @Nonnull
+    default BatchStage<T> distinct() {
+        return distinctBy(wholeItem());
+    }
+
+    /**
+     * Attaches to this stage a stage that emits all the items from this
+     * stage as well as all the items from the supplied stage. The other
```

Comment:
I found the wording here slightly too elaborate, it could be expressed in simpler terms perhaps?